Add initial documentation on issue, PR, and release maintenance (#5)
hapijs/.github: MAINTENANCE.md

# Maintenance
This document outlines basic procedures for maintenance of issues, PRs, and releases.

## Issues and Pull Requests
Issues and pull requests are at the core of hapi's approach to documentation and messaging to the broader community.  They are rigorously labeled, associated with releases, and assigned to collaborators responsible for their resolution.  Official discussions, release notes, announcements, and community support all generally take the form of issues.  Finally, the source of truth for our changelogs are also managed through issues.

### Labels
Labels are standardized through [organization defaults](https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/managing-default-labels-for-repositories-in-your-organization).  Here are some notable ones:
 - `support` - requests for clarification or help.
 - `bug` - reports of unintended behavior.
 - `security` - indication of security implications.
 - `feature` - requests for new functionality.
 - `release notes` - documentation of changes in major releases, sometimes accompanied by a migration guide.

>  - `documentation` - non-code related changes to the readme, API docs, or similar.

Reviewer: ```suggestion
 - `documentation` - non-code related changes to the README, API docs, or similar.
```

>  - `dependency` - changes to underlying or peer dependencies.

Reviewer: I'd just leave out the "underlying or peer" part.

Author: ```suggestion
 - `dependency` - changes to dependencies.
```

 - `breaking changes` - indication of a breaking change in behavior, principally for major releases.
 - `help wanted` - signals that it's welcome for any community member to offer a PR or otherwise help to resolve the issue.

Every issue and pull request should have at least one label attributed to it.  The labels for an issue may change over time.  When a pull request addresses a particular issue, it will often have the same labels as that issue.

### The Connection to Releases

> We associate PRs and issues to released versions of our modules in order to leave a paper trail for users that includes code, documentation, and discussion.  It can also be interpreted programmatically to e.g. display automatically-generated changelogs on the website as can be seen [here for boom](https://hapi.dev/module/boom/changelog/).

Reviewer: ```suggestion
We associate PRs and issues to released versions of our modules in order to leave a paper trail for users that includes code, documentation, and discussion.  It can also be interpreted programmatically to e.g. display automatically generated changelogs on the website as can be seen [here for boom](https://hapi.dev/module/boom/changelog/).
```

#### Milestones

> When a PR is merged or issue is closed, it should be associated with a [milestone](https://github.com/hapijs/hapi/milestones) for the version of the module in which the corresponding work will be released.  All merged PRs should be associated with a milestone, and all issues resolved by a merged PR or a commit should have a milestone.  Milestones are named by the full semver version of the release, e.g. `2.1.0`, `0.1.1`, `9.0.0`.  When the version is published, that milestone should be closed.  There should never be zero open milestones for a given module: when one milestone is closed, a new one is created for the next patch version.  For example, once `2.1.0` is released then the milestone for that version will be closed and a new milestone for `2.1.1` will be created.

Reviewer: It might be worth adding a sentence stating that if things aren't part of a milestone, they won't show up on the website. That was unpleasant to discover the hard way 😄

Author: ```suggestion
When a PR is merged or issue is closed, it should be associated with a [milestone](https://github.com/hapijs/hapi/milestones) for the version of the module in which the corresponding work will be released.  All merged PRs should be associated with a milestone, and all issues resolved by a merged PR or a commit should have a milestone.  Milestones are named by the full semver version of the release, e.g. `2.1.0`, `0.1.1`, `9.0.0`.  When the version is published, that milestone should be closed.  There should never be zero open milestones for a given module: when one milestone is closed, a new one is created for the next patch version.  For example, once `2.1.0` is released then the milestone for that version will be closed and a new milestone for `2.1.1` will be created.

This organization of issues into milestones is central to generating the module changelogs found on [hapi.dev](https://hapi.dev), and to creating transparency around what is contained in each release.
```

#### Release Notes

> It is encouraged to write-up an issue with release notes for major changes, especially for modules that are intended to be used externally to the organization (e.g. hapi, bell, nes) versus hapi core (e.g. pez, topo, shot).  The [hapijs/hapi](https://github.com/hapijs/hapi) repository should have release notes for every major version of the framework.  [This](https://github.com/hapijs/hapi/issues/4017) is a great example of typical release note format and ideal level of detail.

Reviewer: ```suggestion
It is encouraged to create an issue with release notes for major changes, especially for modules that are intended to be used externally to the organization (e.g. hapi, bell, nes) versus hapi core (e.g. pez, topo, shot).  The [hapijs/hapi](https://github.com/hapijs/hapi) repository should have release notes for every major version of the framework.  [This](https://github.com/hapijs/hapi/issues/4017) is a great example of typical release note format and ideal level of detail.
```

### Assignees
The collaborator responsible for final review and merge of pull requests should mark themselves as the "assignee."  This leaves an at-a-glance paper trail primarily for convenience ("who understands the changes made in #365?"), and is in recognition of that person's involvement/contribution.  When a collaborator closes an issue that they were responsible for resolving (e.g. answered a user's support question), they may also mark themselves as the assignee.

### Closing
> _This list is not meant to indicate anything about the numerous other cases that may occur, it is meant only to offer some guidelines._

 - Support issues without any response should remain open for two weeks then may be closed.
 - If a support issue doesn't contain a sufficient reproduction it may be closed immediately with a comment.

>  - Any effort at resolving a support issue makes it eligible for being closed.  It's useful to point folks to the hapi hour slack for further support.

Reviewer: Maybe indicate that discussion can continue on a closed issue.

Author: ```suggestion
 - Any effort at resolving a support issue makes it eligible for being closed.  It's useful to point folks to the hapi hour slack for further support.  It is welcome for discussion to continue on the closed issue.
```

 - `security` - indication of security implications.
 - `feature` - requests for new functionality.
 - `release notes` - documentation of changes in major releases, sometimes accompanied by a migration guide.
 - `documentation` - non-code related changes to the README, API docs, or similar.
 - `dependency` - changes to dependencies.
 - `breaking changes` - indication of a breaking change in behavior, principally for major releases.
 - `help wanted` - signals that it's welcome for any community member to offer a PR or otherwise help to resolve the issue.

Every issue and pull request should have at least one label attributed to it.  The labels for an issue may change over time.  When a pull request addresses a particular issue, it will often have the same labels as that issue.

### The Connection to Releases
We associate PRs and issues to released versions of our modules in order to leave a paper trail for users that includes code, documentation, and discussion.  It can also be interpreted programmatically to e.g. display automatically generated changelogs on the website as can be seen [here for boom](https://hapi.dev/module/boom/changelog/).

#### Milestones

> When a PR is merged or issue is closed, it should be associated with a [milestone](https://github.com/hapijs/hapi/milestones) for the version of the module in which the corresponding work will be released.  All merged PRs should be associated with a milestone, and all issues resolved by a merged PR or a commit should have a milestone.  Milestones are named by the full semver version of the release, e.g. `2.1.0`, `0.1.1`, `9.0.0`.  When the version is published, that milestone should be closed.  There should never be zero open milestones for a given module: when one milestone is closed, a new one is created for the next patch version.  For example, once `2.1.0` is released then the milestone for that version will be closed and a new milestone for `2.1.1` will be created.

Reviewer 2: I never agreed with the rule about always having a milestone opened, that's an unnecessary process and you never know what the next version will be, this should go imho.

Author: I do see where you're coming from.  I pinged Colin and some others to get some quick feedback on this point, and I heard mixed feedback.  For this reason I propose we move forward with this as-is in order to document the current state of things.  One of the most important reasons why I'm writing this documentation is so that we have a living document to work with, allowing us to propose changes such as this one that you're suggesting.

Reviewer 3: I agree that you never know what type of semver change will be the next version but it is a quick way to find what will land in the next version IMO. Also we could also automate this process I believe by taking the latest milestone title and increment the patch version.

Reviewer 2: For me the quick way to know about the next semver (if any) is to look if there's any opened milestone, if none then it's still TBD. But I hear what you're saying, let's keep it that way if that satisfies the most.

### Closing
> _This list is not meant to indicate anything about the numerous other cases that may occur, it is meant only to offer some guidelines._

 - Support issues without any response should remain open for two weeks then may be closed.
 - If a support issue doesn't contain a sufficient reproduction it may be closed immediately with a comment.
 - Any effort at resolving a support issue makes it eligible for being closed.  It's useful to point folks to the hapi hour slack for further support.  It is welcome for discussion to continue on the closed issue.
 - Any issue that doesn't follow the relevant issue template may be closed immediately, optionally with a comment indicating to re-file the issue following the template.
 - Feature requests that will not be satisfied may be closed with comment.
 - Bugs should remain open until resolved.
 - Release notes should be staged prior to publish and closed upon publish.
 - PRs that will not be accepted may be closed with a comment.  If a PR is replaced, the new PR should be referenced in a comment.

## Releases

> When a new version of a module is ready to be published, we adhere to the following release process:

Reviewer 4: This section is clearly targeted at current maintainers, making it more of a guide than a policy. Can we change it to a more generic language to better match the other sections?

Author: I addressed this and your other feedback around the release section.  I left it as a guide, but I tried to make the language a bit more generic.  I also added a short introduction summarizing what a release is and noting the important policies that relate to the integrity of our releases.

 - Ensure the main branch of the repository has passing tests in Travis CI.
 - Clone the repository locally if you haven't already.
 - Navigate to the repository on your machine locally.
 - Ensure you've checked-out the main branch and `git pull`.
 - Ensure there are no changes made locally using `git status`.  There should be no modifications or untracked files.
 - Update `version` in package.json with a new commit marked by an annotated tag using `npm version major|minor|patch`, depending on whether the version contains breaking changes, new features, or only fixes per [semver](https://semver.org/).
 - Push the new commit and the new tag using `git push && git push --tags`.

>  - When you're ready to publish to npm run `npm publish`.  Make sure to have your one-time password handy.

Reviewer 4: I'm missing policy on 2FA. Ie. that maintainers with publish rights are required have npm 2FA enabled, and all releases must be done using 2FA.

>  - When you're ready to publish to npm run `npm publish`.  Make sure to have your one-time password handy.

Reviewer 4: Maybe add a note about publishing patch releases for old branches, to avoid having `latest` point to it.
